docs/sandwich-tests/index.md

```diff
@@ -0,0 +1,478 @@
+
+:::
+
+## Overview
+
+- An opam switch is an isolated OCaml environment
+- Use `opam install` to install third-party libraries into the current opam
+  switch
+- After you install a library, you should:
+  - Add it to your `.opam` file so that it will be installed with all your other
+    dependencies the next time you run `opam install . --deps-only`
+  - Add it to the `libraries` field of your `melange.emit` stanza so your code
+    can use it
+- `melange-fest` is a testing library that allows you write tests in OCaml and
+  run them in Node test runner
+- You can `open` a module to make all its functions available in the current
+  scope
+- Set the value of `module_systems` to `(es6 mjs)` to make Melange generate
+  `.mjs` files that are treated by Node as ECMASCript modules
+- Dune has built-in support for tests
+  - Execute all tests by running `opam exec -- dune build @runtest`
+  - Only re-run succeeding tests when dependencies changed since the last time
+    the test was run
+- The `rule` stanza can be used to specify your tests
+  - Use the `alias` field to attach the rule to the `@runtest` alias
+  - Use the `deps` field to specify the dependencies for the rule
+    - Use variables inside `deps` so you don't have to type a path twice
+  - Use the `action` field to specify the command that launches the test
+
+## Solutions
+
+<b>1.</b> A comprehensive test of `Item.Sandwich.toEmoji` might look like this:
+
+<<< SandwichTests.re#test-to-emoji
+
+<b>2.</b> The `rule` stanza for `BurgerTests.re` is very similar to the existing
+rule we have for `SandwichTests.re`:
+
+```clj
+(rule
+ (alias runtest)
+ (deps
+  Item.re
+  (:input ./output/src/order-confirmation/BurgerTests.mjs))
+ (action
+  (run time node %{input})))
+```
+
+<b>3.</b> A comprehensive test of `Item.Sandwich.toPrice` for Turducken
+sandwiches might look like this:
+
+<<< SandwichTests.re#test-to-price
+
+-----
+
+View [source
+code](https://github.com/melange-re/melange-for-react-devs/blob/main/src/sandwich-tests/)
+and [demo](https://react-book.melange.re/demo/src/sandwich-tests/) for this chapter.
+
+-----
+
+[^1]: The `--deps-only` option tells `opam install` to only install the
+    dependencies and not the package. When you're developing an application and
+    not a library, there is no package associated with your project.
+
+[^2]: In this book, we only use opam's [local
+    switches](https://opam.ocaml.org/blog/opam-local-switches/), which are in
+    practice similar to Node project directories containing a `package.json`
+    file. But in general, an opam switch doesn't have to be associated with a
+    project directory.
```

Comment:
These are called "global" switches, as opposed to "local" (not sure it's worth mentioning)